src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -12,99 +13,322 @@ namespace System.Net.Sockets
-                            }
+                    _currentEngine = new SocketAsyncEngine();
+                }
 
-                            Interop.Sys.SocketEvent* buffer;
-                            err = Interop.Sys.CreateSocketEventBuffer(EventBufferCount, &buffer);
-                            if (err != Interop.Error.SUCCESS)
-                            {
-                                Interop.Sys.CloseSocketEventPort(port);
-                                throw new InternalException();
-                            }
+                engine = _currentEngine;
+                handle = _currentEngine.AllocateHandle(context);
+            }
+        }
 
-                            var engine = new SocketAsyncEngine(port, buffer);
-                            Task.Factory.StartNew(o =>
-                            {
-                                var eng = (SocketAsyncEngine)o;
-                                for (;;)
-                                {
-                                    try
-                                    {
-                                        eng.EventLoop();
-                                    }
-                                    catch (Exception e)
-                                    {
-                                        Debug.Fail(string.Format("Exception thrown from event loop: {0}", e.Message));
-                                    }
-                                }
-                            }, engine, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
-
-                            Volatile.Write(ref _engine, engine);
-                        }
-                    }
+        private IntPtr AllocateHandle(SocketAsyncContext context)
+        {
+            Debug.Assert(Monitor.IsEntered(_lock));
+            Debug.Assert(!IsFull);
+
+            IntPtr handle = _nextHandle;
+            _handleToContextMap.Add(handle, context);
+
+            _nextHandle = IntPtr.Add(_nextHandle, 1);
+            _outstandingHandles = IntPtr.Add(_outstandingHandles, 1);
+
+            Debug.Assert(handle != ShutdownHandle);
+            return handle;
+        }
+
+        private void FreeHandle(IntPtr handle)
+        {
+            Debug.Assert(handle != ShutdownHandle);
+
+            bool shutdownNeeded = false;
+
+            lock (_lock)
+            {
+                if (_handleToContextMap.Remove(handle))
+                {
+                    _outstandingHandles = IntPtr.Subtract(_outstandingHandles, 1);
+
+                    Debug.Assert(_outstandingHandles.ToInt64() >= 0);
+
+                    //
+                    // If we've allocated all possible handles for this instance, and freed them all, then 
+                    // we don't need the event loop any more, and can reclaim resources.
+                    //
+                    if (IsFull && _outstandingHandles == IntPtr.Zero)
+                        shutdownNeeded = true;
```

Comment:
Since we'll basically never hit IsFull in production, this probably doesn't matter, but should we also null out the current engine to allow any state here to be GC'd?